// noren/dist/server/res.js
/* IMPORT */
import Cookie from '../parsers/cookie.js';
import { STATUS } from './constants.js';
import ContentType from '../parsers/content_type.js';
import Headers from './headers.js';
import Logger from './logger.js';
import { isString } from './utils.js';
/* MAIN */
class Res {
    /* CONSTRUCTOR */
    constructor() {
        this.ended = false;
        this.headers = new Headers();
        this.log = Logger;
        this.statusCode = 500;
    }
    /* DETAILS API */
    append(key, value) {
        if (isString(value)) {
            this.headers.append(key, value);
        }
        else {
            value.forEach(value => {
                this.headers.append(key, value);
            });
        }
        return this;
    }
    code(code) {
        this.statusCode = code;
        return this;
    }
    cookie(key, value, options) {
        const cookie = Cookie.stringify(key, value, options);
        this.append('Set-Cookie', cookie);
        return this;
    }
    header(key, value) {
        this.headers.set(key, value);
        return this;
    }
    redirect(url, status = 302) {
        const message = `Redirecting to ${url}...`;
        this.status(status, message);
        this.header('Location', url);
        return this;
    }
    status(code, message = STATUS[code]) {
        this.statusCode = code;
        if (message) {
            this.text(message);
        }
        return this;
    }
    type(type) {
        const contentType = ContentType.get(type);
        this.headers.set('Content-Type', contentType);
        return this;
    }
    /* BODY API */
    html(value) {
        this.type('text/html');
        this.send(value);
        return this;
    }
    json(value) {
        this.type('application/json');
        this.send(JSON.stringify(value));
        return this;
    }
    text(value) {
        this.type('text/plain');
        this.send(value);
        return this;
    }
    send(value) {
        this.body = value;
        return this;
    }
    /* OTHER API */
    end() {
        this.ended = true;
        return this;
    }
}
/* EXPORT */
export default Res;
